### PR TITLE
Refactor/test apps

### PR DESCRIPTION
**Pull request by marinoluck**
- Separate core code from Test (Host) App. 
- Added 2017 Test App
- Added 2018 Test App
- Added 2019 Test App
- Added 2020 Test App
- Added Coverage Tests
- Added About Screen (to easily know the lib version)
- Added Export package script
- Added run_tests.sh script
- Added publish.sh script

### CODE REVIEW

**Review by cristiann91 — DISMISSED** [review]

File: 2019-Container/Packages/packages-lock.json [hunk 1]
```diff
@@ -8,41 +8,67 @@
...
-      "version": "1.1.4",
+      "version": "1.2.1",
       "depth": 0,
       "source": "registry",
       "dependencies": {
         "com.unity.test-framework": "1.1.1"
       },
       "url": "https://packages.unity.com"
     },
+    "com.unity.ide.visualstudio": {
```
Comment (asavizky-sm) [inline 2019-Container/Packages/packages-lock.json:26]: Do we want this?

File: 2019-Container/publich.sh [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+YELLOW='\033[0;33m'
+BRED='\033[1;31m'
+BGREEN='\033[1;32m'
+ENDCOLOR='\033[0m'
+
+export unityPath=/Applications/Unity/Hub/Editor/2019.4.28f1
+echo '\n'
+echo "Exporting package, please wait ...\n"
+$unityPath'/Unity.app/Contents/MacOS/Unity' -batchmode -quit -executeMethod ScaleMonk.Ads.ExportScaleMonk.Export
+# result=$(xmllint --xpath "string(//test-run/@result)" test-results.xml)
```
Comment (asavizky-sm) [inline 2019-Container/publich.sh:14]: Seems like this file name has a typo. Also, do we need all these commented lines?

File: 2019-Container/results.xml [hunk 3]
```diff
@@ -0,0 +1,114 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (asavizky-sm) [inline 2019-Container/results.xml:1]: Should we add this file to gitignore?

File: 2019-Container/test-results.xml [hunk 4]
```diff
@@ -0,0 +1,138 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (asavizky-sm) [inline 2019-Container/test-results.xml:1]: Same with this file

File: SourceTests/AnalyticsTests.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            adsBinding.Received(1).CreateAnalyticsBinding();
+        }
+
+        private static Dictionary<string, string> ContainsEventParam(string key, string value)
+        {
+            return Arg.Is<Dictionary<string, string>>(parameters =>
+                parameters.ContainsKey(key) && parameters[key] == value);
+        }
+    }
+}
```
Comment (asavizky-sm) [inline SourceTests/AnalyticsTests.cs:109]: nit: extra line

File: SourceTests/UserTypeTests.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+            scaleMonkAds.Initialize(() => { });
+            
+            // When set custom user Id 
+            scaleMonkAds.SetUserType(UserType.NON_PAYING_USER);
+
+            // Then the custom user Id is not set
+            adsBinding.Received().SetUserType(UserType.NON_PAYING_USER);
+        }
+    }
+}
```
Comment (asavizky-sm) [inline SourceTests/UserTypeTests.cs:76]: nit: extra line

File: SourceTests/InitializationTests.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+            // When we call the initialization of the SDK two times
+            scaleMonkAds.Initialize(() => { });
+            scaleMonkAds.Initialize(() => { });
+
+            // Then it's initialized just once
+            adsBinding.Received(1).Initialize(scaleMonkAds);
+        }
+    }
+}
```
Comment (asavizky-sm) [inline SourceTests/InitializationTests.cs:49]: nit: extra line

File: SourceTests/CustomUserIdTest.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            {
+                // When set custom user Id 
+                scaleMonkAds.SetCustomUserId("userId");
+            });
+
+            // Then the custom user Id is not set
+            adsBinding.Received(1).SetCustomUserId("userId");
+        }
+    }
+}
```
Comment (asavizky-sm) [inline SourceTests/CustomUserIdTest.cs:47]: nit: extra line

File: SourceTests/AndroidBindingTests.cs [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+
+            // Then the java bridge is called with the right parameters
+            _mockAndroidJavaBridge.Received(1).CallNativeMethod(
+                "setUserType",
+                Arg.Is<object[]>(parameters =>
+                    parameters[0].ToString() == "non_paying_user")
+            );
+        }
+    }
+}
```
Comment (asavizky-sm) [inline SourceTests/AndroidBindingTests.cs:132]: nit: extra line

File: Source/Source/ScaleMonkAds.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+        /// Initialize ScaleMonk SDK
+        ///
+        /// <param name="callback">The callback that will be called after the ScaleMonk SDK is initialized</param>
+        /// </summary>
+        public static void Initialize(Action callback)
+        {
+            SharedInstance.Initialize(callback);
+        }
+    }
+}
```
Comment (asavizky-sm) [inline Source/Source/ScaleMonkAds.cs:53]: nit: extra line

File: Source/Source/Android/AndroidJavaBridge.cs [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+            _adsBinding.Call(methodName, _activity, args);
+        }
+
+        public bool CallBooleanNativeMethod(string methodName, params object[] args)
+        {
+            return _adsBinding.Call<bool>(methodName, _activity, args);
+        }
+    }
+}
+#endif
```
Comment (asavizky-sm) [inline Source/Source/Android/AndroidJavaBridge.cs:42]: nit: extra line

File: Source/Source/Android/AdsAndroidBinding.cs [hunk 12]
```diff
@@ -73,23 +71,32 @@ public bool AreInterstitialsEnabled()
...
+        
+        public void SetUserType(UserType userType)
+        {
+            _androidJavaBridge.CallNativeMethod("setUserType", userType.ToStringUserType());
         }
     }
-}
-#endif
\ No newline at end of file
+}
```
Comment (asavizky-sm) [inline Source/Source/Android/AdsAndroidBinding.cs:102]: nit: extra line

File: Source/Source/Android/AdsAndroidBinding.cs [hunk 13]
```diff
@@ -5,39 +5,37 @@
 // https://www.scalemonk.com/legal/en-US/mediation-license-agreement/index.html 
 //
 
-#if UNITY_ANDROID
```
Comment (asavizky-sm) [inline Source/Source/Android/AdsAndroidBinding.cs:8]: Why you removed this if? Should we compile it even if we are on iOS?

File: Source/Source/AdsFactory.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    public static INativeBridgeService NativeBridgeService()
+    {
+        return new DefaultNativeBridgeService();
+    }
+
+    public static AnalyticsService AnalyticsService()
+    {
+        return new AnalyticsService();
+    }
+}
```
Comment (asavizky-sm) [inline Source/Source/AdsFactory.cs:33]: nit: extra line

File: Source/Editor/Export/ExportScaleMonk.cs [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+                exportedPackageAssetList.Add(path);
+            }
+
+            //Export Shaders and Prefabs with their dependencies into a .unitypackage
+            AssetDatabase.ExportPackage(exportedPackageAssetList.ToArray(),
+                "ScaleMonk" + Config.Version + ".unitypackage",
+                ExportPackageOptions.Recurse | ExportPackageOptions.IncludeDependencies);
+        }
+    }
+}
```
Comment (asavizky-sm) [inline Source/Editor/Export/ExportScaleMonk.cs:27]: nit: extra line

File: Source/Editor/Export/ExportPackage.cs [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+{
+    public class ExportPackage : MonoBehaviour
+    {
+        [MenuItem("ScaleMonk/Export")]
+        static void Export()
+        {
+            ExportScaleMonk.Export();
+        }
+    }
+}
```
Comment (asavizky-sm) [inline Source/Editor/Export/ExportPackage.cs:14]: nit: extra line

**Review by asavizky-sm — COMMENTED** [review]
Can you please add a screenshot of the new adnets window?

File: 2019-Container/Packages/packages-lock.json (hunk 1 above)
Comment (marinoluck) [reply 2019-Container/Packages/packages-lock.json:26]: this is something visual studio add to the testapp

**Review by nravasi — APPROVED** [review]

**Review by nravasi — APPROVED** [review]

**Pull request merged**